pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -132,6 +145,178 @@ func storageMatches(set *apps.StatefulSet, pod *v1.Pod) bool {
 	return true
 }
 
+// getPersistentVolumeClaimPolicy returns the PVC policy for a StatefulSet, returning a retain policy if the set policy is nil.
+func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.StatefulSetPersistentVolumeClaimRetentionPolicy {
+	policy := apps.StatefulSetPersistentVolumeClaimRetentionPolicy{
+		WhenDeleted: apps.RetainPersistentVolumeClaimRetentionPolicyType,
+		WhenScaled:  apps.RetainPersistentVolumeClaimRetentionPolicyType,
+	}
+	if set.Spec.PersistentVolumeClaimRetentionPolicy != nil {
+		policy = *set.Spec.PersistentVolumeClaimRetentionPolicy
+	}
+	return policy
+}
+
+// claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
+// PVC deletion policy for the StatefulSet.
+func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	policy := set.Spec.PersistentVolumeClaimRetentionPolicy
+	const retain = apps.RetainPersistentVolumeClaimRetentionPolicyType
+	const delete = apps.DeletePersistentVolumeClaimRetentionPolicyType
+	switch {
+	default:
+		klog.Errorf("Unknown policy %v; treating as Retain", set.Spec.PersistentVolumeClaimRetentionPolicy)
+		fallthrough
+	case policy.WhenScaled == retain && policy.WhenDeleted == retain:
+		if hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case policy.WhenScaled == retain && policy.WhenDeleted == delete:
+		if !hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case policy.WhenScaled == delete && policy.WhenDeleted == retain:
+		if hasOwnerRef(claim, set) {
+			return false
+		}
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	case policy.WhenScaled == delete && policy.WhenDeleted == delete:
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		// If a pod is scaled down, there should be no set ref and a pod ref;
+		// if the pod is not scaled down it's the other way around.
+		if podScaledDown == hasOwnerRef(claim, set) {
+			return false
+		}
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	}
+	return true
+}
+
+// updateClaimOwnerRefForSetAndPod updates the ownerRefs for the claim according to the deletion policy of
+// the StatefulSet. Returns true if the claim was changed and should be updated and false otherwise.
+func updateClaimOwnerRefForSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	needsUpdate := false
+	// Sometimes the version and kind are not set {pod,set}.TypeMeta. These are necessary for the ownerRef.
```

Comment:
Do you see this issue only in unit tests? Setting the type and object meta in the controller seems odd.

Reply:
The unit tests, and the e2e tests IIRC. It seems to be set incompletely and then gets filled in by the API server? I don't understand myself.

Reply:
The tests should use the object returned from the Create() call so it has all the proper meta filled in. That may be the issue?

Reply:
In the e2e test I have to get the pod name, because it isn't always in the ownerRef. So I'm not really sure of the sequence of when that gets filled in.

The tests use CreatePod from the client, although maybe because it's a testing client something is different?

Reply:
I had a chance to double-check this issue.

It does happen during the e2e tests. For example, from [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L477) the replica does not have the type meta set.

Not clear why, if we've gotten that far down in the controller the pod has already been created. Surely DeepCopy would pick up the type meta?

Reply:
Ok it looks like that replica Pod is just generated from [in-code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L549). It's not reading from the API server.

It looks like it's an existing issue, so I don't think we should block on this, but may be worth trying to refactor in the future. I think setting the meta where it's generated looks better vs when we read it.

Reply:
Not really relevant to this PR, but dropping a breadcrumb for my future self: I'm not sure it's only the create from template, after adding some logging it appears the pod list here never has any typemeta set. Perhaps it is lost in some of the preprocessing (eg [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set.go#L303)). So it might be a larger refactor.

Reply:
objects read from the API server do not always have kind/apiVersion set. This occurs when listing objects. See https://github.com/kubernetes/kubernetes/issues/3030

Are you relying on this always being populated?

Reply:
I fill them in when setting an ownerRef. I'm not relying on them being populated, @msau42 was wondering if my code to fill them in was necessary.